README.md
# pwalk

This repository contains a high-speed multi-threaded treewalk utility called 'PowerWalk' (pwalk)
for various ends including metadata extraction and formatting, tactical filesystem repair work,
and utility operations such as bulk file deletion.  Also included is an assortment of ancillary
programs and documentation.

These programs are all by Bob Sneed of Dell EMC (Bob.Sneed@Dell.com) as part of a collection of
utilities sometimes collectively referred to as "Bobware". While some of the utilities in the
Bobware collection date back to the early 1980's, pwalk has only been under development since 2012.
It has accumulated a unique set of features for solving various problems in a scale-out NAS
environments.

pwalk is an ongoing work-in-progress, with the ambition of becoming an increasingly general-purpose
tool for various operations on highly-scaled filesystems, including but not limited to;

	- high-speed parameter-driven metadata extraction and formatting
	- bulk data operations (e.g. bulk file comparisons)
	- data migration aid (e.g. POSIX-to-NFS4 ACL migration)
	- file heirarchy comparisons with -cmp
	- being a framework for custom-coded tasks requiring a high-speed treewalker foundation

pwalk code management was moved to Github as a private project on September 29, 2017, and later
made public.  It is not intended that anyone come into possession of this code without first
collaborating with the author, Bob Sneed.

*** DISCLAIMERS ***

1. This is FREE CODE. Appropriate recognition of the author in derivative works
	is both ethical and appreciated, but not legally mandated.
2. This code carries no warrantees of any sort whatsoever, including any implied
	warrantees of correctness or suitability for any particular purpose.
3. This code is not a supported product of Dell EMC or any other commercial entity.
4. Apart from the provided documentation, the code is the documentation.
5. In no event shall the author be liable in any way for any damages that may arise
	from using this code or derivatives of this code.
6. The deployment and use of any code associated herewith under Isilon OneFS is
	neither condoned nor supported by Dell EMC.
7. Use at your own risk.

*** All That Being Said ... ***

Documentation for pwalk is spread across a number of documents;

	- pwalk_slides.pptx (Powerpoint) - the best basic orientation to pwalk
	- pwalk_acls.docx (Word) - detailed discussion of ACL-handling features
		and associated utility programs
	- pwalk_manpage.docx (Word) - preliminary manpage (VERY prelilminary)
	- acl_decoder.xlsx (Excel) - 'secret decoder ring' for ACL permissions

Both the code and the documentation are maintained in a perpetual state of incompleteness
and begging for some cleanup and consolidation work!

*** Older Distribution Method ***

pwalk is often distributed as an E-Mail attachment in a uuencoded gzip'ed tarball format
('bundle') which has been designed to get past most E-Mail filters. The C-shell script
'uu-export.csh' creates these bundles in the directory superior to this directory, with
a name than includes a timestamp of when the bundle was created.

To unpack this format, the recipient should take the following steps;

	- Save the E-Mail attachment to a file, e.g. 'pwalk-export_20180312@161642_uu.allow'
	- Edit the file to remove the first line
	- uudecode -p < EDITTED_FILE | tar xzvf  -
		o  This will create a directory called pwalk with about 16 MB of content
		o  pwalk/bin/* will contain the last-built pwalk-related binaries

NOTE: The various bin/* directories are NOT assured to be in-sync from the same source code!

*** ERRATA ***

NOTICE: 2.05b3 was incorrectly committed as 2.05b8 (8 versus 3).  The next push will  ***
will end with b4, and the b8 anomally will live in infamy. This build purports to     ***
correct major brokeness in the ACL-related code and related utilities.                ***



src/CHANGELOG
Version.FUTURE [ DEVELOPMENTAL TO-DO ]
	ADD: Copy -L -D -H symlink-handling options from 'du'.
	SOMEHOW: Do something to avoid multiple-processing of hard-linked inodes
	$$$$ Add find(1)-like selection predicates and actions
	Add external parameterization for -csv[=<list>] and [csv] pfile section (report column keys)
		Inspired by: ps '-o' option (inode,asize,mode,nlink,nsize,name)
	Add external parameterization for -select in [tally] pfile section (file selection criteria)
	Add external parameterization for +tally in [tally] section of parameter file
	Add MD5 checksum features to -cmp mode
	Make concurrency runtime-dynamic to allow varying impact of long-running jobs
	Improve periodic status updates; make on-demand, add moving average, estimated current files/sec
	- FIX: Always emit asccify'd directory and file names in primary outputs
	- NEW: -select=fake (OneFS-only) which uses aclu_get_sd() in pwalk_onefs.c
	- $$ Add pclt2 metadata items on OneFS
		... Add ifs.file_pool metadata (OneFS only)
		... Remove pwalk_python.py dependency for -audit
	- Use /dev/papi ? to fetch WORM data rather than syscall
		... might later go over http ...
	- Add -trash primary mode to move selected files to [target] (with mkdir -p if needed)
...
Version 2.10 - 2020/07 - New features & fixes ...
	- NEW: -select_regex=<regex> - filenames matching <regex>, case-insensitive, extended syntax
	- NEW: -select=sparse - files which appear to be sparse (DEVELOPMENTAL)
	- NEW: -select=fake - files with uid or gid in 1000000-4000000 range (DEVELOPMENTAL)
		When used with -ls, uid and gid are added to output
		MISSING: call to determine actual PERSISTED IDs
	- NEW: Selection options are now implicitly AND-ed!
	- FIX: When any -select option used, do not include directory sizes in size statistics,
		in other words, exclude them from per-directory subtotals and +tally statistics.
	- FIX: +tally Inflation% heading corrected; now just "Inflation"
Version 2.09b5 - 2020/07/08 - Bugfix ...
	- FIX: fixed longstanding macOS issue with -gz pipe closing (I hope)
Version 2.09b4 - 2020/04 - Minor bugfix ...
Version 2.09b3 - 2020/04 - New features & fixes ...
	- FIX: hacls now shows OneFS "Owner Rights" as "owner_rights"
	- NEW: xacls -s1c and -s1l show OneFS chmod and 'ls -l' respectively
	- FIX: +tally horribly broken in 2.09b2
	- NEW: +tally now has [tally] section in pfile for custom reports
Version 2.09b2 - 2020/04 - New features & fixes ...
	- NEW: +tally bucket sizes changed to a superset of bucket sizes reported by isi_distill
Version 2.09b1 - 2020/03 - New features & fixes ...
	- NEW: -select=lfn to select files with long (> 255 byte) filenames
	- NEW: -[not]since*= options invert select sense
	- NEW: In pwalk.log, files *scanned* are reported distinctly from files *selected*
Version 2.08 - 2019/11 - New features & fixes ...
	- FIX: -redact now always use hex-formatted inode numbers (files were previously decimal)
	- FIX: Do not open symlinks under OneFS! (oops in dirent readonly ops)
	- FIX: Restore Python-based -audit functionality (alpha code is Python-less)
	- NEW: -select=[no]stubs (OneFS-only) which uses st_flags bit
	- NEW: -select_[atime|mtime|ctime|birth]=<ref_time>
	- NEW: -lsf ('full') shows full pathname for every dirent output (ie: 'selected')
	- pwalk -audit logic now properly isolated in pwalk_audit.[ch]
	- Changed nomenclature to 'logical' and 'physical' (lsize and psize) from 'nominal' and 'allocated'
	- On OneFS, '+' always included in mode bits when ACL present (it's just a flag bit!)
	- -lsc corrected to *not* emit inode numbers
	- Skips (eg: .snapshot, .ifsvar) now log to pwalk.log rather than worker*.* files
	- mystat now emits sb.st_flags keywords for non-zero st_flags (useful for seeing stubs, etc)
Version 2.07 - 2019/02 - New features ...
	- Option -since=<file> now implicitly enables -select
	- Added +span option to span filesystems (opposite to find(1) -xdev) (based on stat(2) st_dev)
	- Added tracking of largest inode number 'seen' and 'selected'
	- Fixed -gz on OSX by using fflush() plus fclose() on pipes instead of pclose()
	- Fixed new threading model on OSX by correcting worker mutex initializations
	- New thread management has no polling loop, and shows work time more accurately
	- Re-structured thread management to solve hang-before-exit race conditions
	- Bugfix: all pwalk outputs now owned by pwalk's user rather than setuid/setgid identity
Version 2.06 - 2018/09 - 2019/01 - New Features in-process
	- Added "tolerate '\n' and other binary in FIFO pathnames" (asciify())
	- Bugfix; .err outputs now line-buffered (was erroneously full-buffered)
	- Default <directory> is now "." (relative to SOURCE) if none given on command line
	- Bugfix: "Hang on starting" - streamlined directory_scan() & eliminated WAKEUP mutex
	- Added read and write of NTFS ACLs (NFS4, OSX-native, OneFS-native, and possibly Windows-native)
	- Added +md5 and +sha1 options (IN-PROGRESS; calculate using same f() as +crc!)
		... optionally convey in wacls data stream
		... store md5 and/or sha1 checksums as an attribute during -cp operations
		... validate calculated versus stored md5 checksums
		... store md5 as attribute on target with -cp +md5
	- Added -cp primary mode to mass copy files to [target] (data, owner, group, sd)
		... where the sd (security descriptor) can be processed in-band
	- Updated wacls: convey/capture additional metadata
		... add chown, chgrp, chmod, utimes, md5, sha1, (OneFS futimes & WORM state)
		... v2 stream format, preserving compatabllity with v1.0 format
		... Unity-native wacls build
...
Version 2.06b1 - 2018/10/17 - New Features & changed behaviors
	- Updated slide deck
	- Added rudimentary DS/WS/GS counters for hard links & zero-length files
	- Added -redact option; output only hex inode numbers for most modes
	- Added -bs=512 option; default changed to 1024
	- +tally output is now 'pwalk_tally.csv' and hard-coded for file size buckets

src/pwalk.h
#ifndef PWALK_H
#define PWALK_H 1

#include <sys/param.h>

// @@@ Portabiity tidbits ...
// See also: http://sourceforge.net/p/predef/wiki/OperatingSystems/

typedef unsigned long long count_64;
typedef enum {EMBRYONIC=0, IDLE, BUSY} wstatus_t;	// Worker operational status

#if !defined(TRUE)
#define TRUE 1
#endif
#if !defined(FALSE)
#define FALSE 0
#endif

#if __APPLE__ && __MACH__
#define __OSX__ 1
#include <mach/mach.h>
#include <mach/mach_time.h>
#endif

#if defined(LINUX)
#define __LINUX__ 1
#define HAVE_STRUCT_STAT_ST_FLAGS 0
#else
#define HAVE_STRUCT_STAT_ST_FLAGS 1
#endif

#if (defined(__FreeBSD__) || defined(__OSX__)) && !defined(BSD)
#define BSD 1
#endif

#if defined(BSD) || defined(__LINUX__)
#define BSDLINUX 1
#endif

// @@@ Platform dependencies ...

#if defined(SOLARIS)
#define PWALK_PLATFORM "Solaris"
#elif defined(__LINUX__)
#define PWALK_PLATFORM "Linux"
#elif defined(__OSX__)
#define PWALK_PLATFORM "OSX"
#elif defined(__ONEFS__)
#define PWALK_PLATFORM "OneFS"
#else
#define PWALK_PLATFORM "?Unknown Platform?"
#endif

#if defined(__LINUX__)
#define PWALK_ACLS 1		// NOTE: Some ACL-related code is merely dormant or innocuous
static int CLK_TCK;
#if !defined(PTHREAD_MUTEX_ERRORCHECK)
#define PTHREAD_MUTEX_ERRORCHECK PTHREAD_MUTEX_ERRORCHECK_NP
#endif // !defined(PTHREAD_MUTEX_ERRORCHECK)
#if !defined(CLOCK_MONOTONIC_RAW)
#define CLOCK_MONOTONIC_RAW CLOCK_MONOTONIC
#endif // !defined(CLOCK_MONOTONIC_RAW)
#ifdef TIMESPEC_HACK // klooge: only OLD Linux
#define st_atimespec st_atim
#define st_mtimespec st_mtim
#define st_ctimespec st_ctim
#define st_birthtimespec st_ctim
#define st_birthtime st_ctime
#endif // TIMESPEC_HACK
#else // -> !defined(__LINUX__)
#define PWALK_ACLS 0
#endif // defined(__LINUX__)

#if defined(__ONEFS__)
#define PWALK_AUDIT 1		// PWALK_AUDIT enables OneFS-only SmartLock audit functionality
#define USE_VTIMES 1
#include <sys/extattr.h>
#include <sys/isi_enc.h>	// for real lvtimes()/vtimes() (OneFS private)
#include <ifs/ifs_types.h>
#include <ifs/ifs_syscalls.h>
#include <ifs/bam_pctl.h>
#else				// use 'stub' versions of these to reduce inline conditional code
#define O_OPENLINK 0		// klooge: also for native BSD?
#define PWALK_AUDIT 0
#define USE_VTIMES 0
#include <sys/xattr.h>
#define VT_ATIME 1
#define VT_MTIME 2
#define VT_BTIME 4
#endif // defined(__ONEFS__)

// @@@ Global defines and shared global data ...

#define MAX_WORKERS 128				// Max -dop= value
#define MAX_MKDIR_RETRIES 32			// ????
#define MAXPATHS 64				// Arbitrary limit for [source] or [target] multi-paths
#define MAX_PATH_DEPTH 128			// Max pathname components
#define PROGRESS_TIME_INTERVAL 3600/4		// Seconds between progress outputs to log file

// Mask bits for metadata to gather during treewalk (in PWget_MASK) ...
#define PWget_STAT	0x001		// Basic stat()
#define PWget_OWNER	0x002        	// Owner name
#define PWget_GROUP	0x004        	// Group name
#define PWget_STUB	0x008		// (not implemented)
#define PWget_ACLP	0x010		// Linux only: POSIX ACL
#define PWget_ACL4	0x020		// NFS4 ACL (not implemented)
#define PWget_WORM	0x040		// OneFS only: WORM state
#define PWget_SD	0x080        	// OneFS only: Security Descriptor

typedef struct {
   void *ptr;		// point to area
   unsigned size;	// size of area
   void *copy;		// known-good copy of area
   char msg[256];	// utility buffer
} PWALK_DEBUG_BLOCK;

#if defined(PWALK_SOURCE)
FILE *Plog = NULL;			// Main logfile output (pwalk.log)
int PWget_MASK = PWget_STAT;		// Mask of PWget_* bits (always stat(), at least)
int VERBOSE = 0;			// -v (verbosity) 1, 2, 3, ... more and more verbose
int PWdebug = 0;			// -d (debug) level
int PWquiet = 0;			// -q (quiet) flag
int PWdryrun = 0;			// -dryrun flag
PWALK_DEBUG_BLOCK PBLK;
#else
extern FILE *Plog;
extern int PWget_MASK;
extern int VERBOSE;
extern int PWdebug;
extern int PWquiet;
extern int PWdryrun;
extern void abend(char *str);
extern PWALK_DEBUG_BLOCK PBLK;
#endif

// @@@ Global parameters for +tally  ...

#define MAX_TALLY_BUCKETS 128                   // MUST be a #define!
static int N_TALLY_BUCKETS;			// Counted at runtime for TALLY_BUCKET_SIZE array

// -tally scoreboard block (only for WS and GS, not DS) ...
typedef struct {
   count_64 count[MAX_TALLY_BUCKETS];
   count_64 size[MAX_TALLY_BUCKETS];
   count_64 space[MAX_TALLY_BUCKETS];
} TALLY_BUCKET_COUNTERS;

static char *TALLY_TAG = "tally";               // Default '+tally=<tag>' value
static char *TALLY_COLUMN_HEADING[] = {
   "Tag","Bucket","Count","Count%","sum(Size)","Size%","sum(Space)","Space%","Inflation",NULL
};
static count_64 TALLY_BUCKET_SIZE[MAX_TALLY_BUCKETS] = {
   0,
   1*1024,		// KiB
   2*1024,
   3*1024,
   4*1024,
   5*1024,
   6*1024,
   7*1024,
   8*1024,
   16*1024,
   24*1024,
   32*1024,
   40*1024,
   48*1024,
   56*1024,
   64*1024,
   72*1024,
   80*1024,
   88*1024,
   96*1024,
   104*1024,
   112*1024,
   120*1024,
   128*1024,
   256*1024,
   512*1024,
   1048576,		// MiB = 1024^2 = 2^20
   10*1048576,
   100*1048576,
   (1L<<30),		// GiB = 1024^3 = 2^30
   10*(1L<<30),
   100*(1L<<30),
   (1L<<40),		// TiB = 1024^4 = 2^40
   0
};

// @@@ WorkerData is array of structures that contains most worker-indexed private DATA ...
#if defined(PWALK_SOURCE)
struct {                         	// Thread-specific data (WorkerData[i]) ...
#else
extern struct {
#endif
   // Worker-related ...
   int                  w_id;			// Worker's unique index
   wstatus_t            status;			// Worker status
   FILE                 *wlog;			// WLOG output file for this worker
   FILE                 *werr;			// WERR output file for this worker
   // Co-process & xacls support ...
   FILE                 *PYTHON_PIPE;		// Pipe for -audit Python symbiont
   FILE                 *WACLS_PIPE;		// Pipe for +wacls= process
   FILE                 *XACLS_BIN_FILE;	// File for +xacls=bin output
   FILE                 *XACLS_CHEX_FILE;	// File for +xacls=chex output
   FILE                 *XACLS_NFS_FILE;	// File for +xacls=nfs output
   FILE                 *XACLS_ONEFS_FILE;	// File for +xacls=onefs output
   // Pointers to runtime-allocated buffers ...
   char                 *DirPath;		// Fully-qualified directory to process
   struct dirent        *Dirent;		// Buffer for readdir_r()
   void                 *SOURCE_BUF_P;		// For -cmp source
   void                 *TARGET_BUF_P;		// For -cmp source
} WorkerData[MAX_WORKERS+1];			// klooge: s/b dynamically-allocated f(N_WORKERS) */

// @@@ Statistics blocks ...
//
// Statistics are generaly collected in three phases to avoid locking operations;
//	1. At the per-directory level (DS) - during each directory scan
//		- We need per-directory subtotals in some outputs (but not for +tally)
//	2. At the per-worker level (WS) - sub-totaled at the end of each directory scan
//		- We do not want lock competition between workers while scanning
//	3. At the pwalk global level (GS) - grand-total summed from worker subtotals at the very end
//		- Aggregating the grand totals is a lock-less operation because all workers are done

// Statistics: cascades from directory (DS) to worker (WS) to grand total (GS) summary ...
typedef struct {
   // Accumulated per-directory ...
   count_64 NOpendirs;				// Number of opendir() calls
   count_64 NScanned;				// Files scanned (superset of selected files)
   count_64 NSelected;				// Files selected (when selection option(s) given)
   count_64 NRemoved;				// Files removed (with -rm)
   count_64 NACLs;				// +acls, +xacls=, or +wacls= # files & dirs w/ ACL processed
   count_64 NStatCalls;				// Number of lstatat() calls on dirents
   count_64 NDirs;				// ... # that were directories
   count_64 NFiles;				// ... # that were files
   count_64 NSymlinks;				// ... # that were symlinks
   count_64 NOthers;				// ... # that were others
   count_64 NStatErrs;				// ... # that were errors
   count_64 NWarnings;				// Scan issues other than stat() failures
   count_64 NZeroFiles;				// Number of ordinary files of size 0
   count_64 NHardLinkFiles;			// Number of non-directories with link count > 1
   count_64 NHardLinks;				// Sum of hard link counts > 1
   off_t NBytesPhysical;			// Sum of allocated space
   off_t NBytesLogical;				// Sum of nominal file sizes
   // Accumulated per-worker for selected() files ...
   count_64 READONLY_Zero_Files;		// READONLY zero-length files
   count_64 READONLY_Opens;			// READONLY file opens
   count_64 READONLY_Errors;			// READONLY open/read errors
   count_64 READONLY_CRC_Bytes;			// READONLY CRC bytes read
   count_64 READONLY_DENIST_Bytes;		// READONLY DENIST bytes read
   count_64 NPythonCalls;			// Python calls
   count_64 NPythonErrors;			// Python errors
   count_64 MAX_inode_Value_Seen;		// Cheap-to-keep (WS, GS) stats
   count_64 MAX_inode_Value_Selected;
   TALLY_BUCKET_COUNTERS TALLY_BUCKET;		// +tally counters
} PWALK_STATS_T;

#if defined(PWALK_SOURCE)
PWALK_STATS_T GS;				// 'GS' is 'Global Stats', only referenced in main code
PWALK_STATS_T *WS[MAX_WORKERS+1];		// 'WS' is 'Worker Stats', calloc'd (per-worker)
char *PYTHON_COMMAND = NULL;			// For OneFS -audit operation
#else
extern PWALK_STATS_T GS;
extern PWALK_STATS_T *WS[];
extern char *PYTHON_COMMAND;
#endif

// Convenience MACROs for worker's thread-specific values ...
// ... so 'fprintf(WLOG' becomes 'fprintf(WorkerData[w_id].wlog' ...
// ... which of course requires a context in which 'w_id' is defined.
#define WDAT WorkerData[w_id]		// Coding convenience for w_id's worker data
#define WLOG WDAT.wlog			// Coding convenience for w_id's output FILE*
// Per-worker .err files get created only when needed ...
#define WERR (WDAT.werr ? WDAT.werr : worker_err_create(w_id))  // Coding convenience for w_id's error FILE*

#endif
